### PR TITLE
Split `release` profile into `release` and `release-opt`

### LINKED ISSUE #1477: Unclear intent regarding optimizations for some profiles
### Current behavior 😯

There are a few conditions related to build profiles for `gitoxide` defined in the top-level `Cargo.toml` that seem like they can be addressed when—as planned in the comment discussion in #1475—splitting the `release` profile into separate `release` and `prof` (or `profile` or `profiling`) profiles.

1. [ ] Debug symbols - as discussed.
2. [ ] Unclear intent for LTO - `#lto = "fat"`.
3. [ ] Old plan for optimizations in debug builds.

#### 1. Debug symbols

Currently the `release` profile leaves debug symbols in because they are useful for profiling, but this creates the need to strip them in a separate step when they are not wanted, such as when building binaries to publish in releases on GitHub.

This is as discussed in comments in #1475, with a suggested plan in https://github.com/Byron/gitoxide/pull/1475#issuecomment-2256168338.

#### 2. Unclear intent for LTO

It is not clear if [link-time optimization](https://doc.rust-lang.org/cargo/reference/profiles.html#lto), at higher than the default very limited level, should be done. It's possible this might even differ between `release` and `prof` profiles. (This is the main sub-issue that may need examination and discussion, and the main reason I opened this issue.)

Early on, in 07859cb, various performance-related configuration [was set](https://github.com/Byron/gitoxide/commit/07859cb6031ccc1938f66481ec1dd88cf05185b8#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542) for the `release` profile. This included the most aggressive link-time optimization available—short of [LTO across different languages](https://doc.rust-lang.org/rustc/linker-plugin-lto.html)—which was still in place about a year ago:

https://github.com/Byron/gitoxide/blob/71efcbba112376b4acaf37d662cdb38d369462be/Cargo.toml#L185

Then in bd32e39 (#893) it was commented out, with the effect of changing the `lto` setting from `fat` to [the default](https://doc.rust-lang.org/cargo/reference/profiles.html#release) of `false`, which somewhat unintuitively still performs link-time optimization, just only within the crate itself.

https://github.com/Byron/gitoxide/blob/1d37bf6a773d56eea9003aa626ced413e8e0eaa3/Cargo.toml#L209

For the last year, it's been that way. Since the `gitoxide` project is made up of many crates, which the `gitoxide` binary crate itself uses, I suspect this does not leverage most of the benefit of LTO. That is, the word "`false`" may be intuitively accurate in this case.

Perhaps benchmarking was done and found LTO not to be helpful for `gix` and `ein`. (In other software that uses any `gitoxide` crates, that software would of course set whatever LTO-related configuration is wanted.) But the commit and PR do not state a reason for the change, and they are otherwise not conceptually related to LTO or optimization. If the goal is to decrease *build* times, then `thin` could be used, which is documented to be almost as good as `fat` while being faster to perform.

Another possibility is that LTO was turned off because the resulting optimizations decreased the detail available when profiling or made the results hard to interpret. If that was the reason, then I think that, when splitting `release` into `release` and `prof`, this would be resolved by setting different values of `lto` in the two different profiles. `release` could have `fat` or `thin`, while `prof` could stick with the default of `false`.

#### 3. Old plan for optimizations even in debug builds?

This would not affect the `release` or `prof` configuration, but I suspect it could be clarified at the same time.

a5b1cd3 added:

https://github.com/Byron/gitoxide/blob/1d37bf6a773d56eea9003aa626ced413e8e0eaa3/Cargo.toml#L217-L219

It seems intuitively odd to compile dependencies with a higher optimization level than the code that is using them, especially since this may make stack traces less rich. On the other hand, maybe it is to make test runs faster.

We are already doing it, with the even more aggressive level of `3`, in the `dev` profile for particular dependencies, most of which are even `gix-*` crates. This was done in 4a948d0 (after having been attempted in #190 per a22c95b and 85f1a48) with the explicit purpose of making some tests faster. I am not suggesting that *this* be changed, since its purpose is clear from the commit message and running tests faster has a number of benefits.

https://github.com/Byron/gitoxide/blob/1d37bf6a773d56eea9003aa626ced413e8e0eaa3/Cargo.toml#L196-L205

But one would expect that and the commented-out `opt-level = 2` for `*` to be right next to each other. Instead, the `release` profile is written in between them. Although the commit message in a5b1cd3 makes clear that this was not why that commented-out code was introduced, I wonder if part of the reason it has been preserved as a comment for so long, rather than eventually being uncommented or removed, could be confusion with this other code in the `release` configuration:

https://github.com/Byron/gitoxide/blob/1d37bf6a773d56eea9003aa626ced413e8e0eaa3/Cargo.toml#L215

To be clear, I'm not at all advocating changing *that*, which was introduced in add2e3e (#2) and makes it so dependencies that are only used when building and not used when running the code are not needlessly optimized.

When adding a `prof` profile, it would go right next to the `release` profile, and having both of them split the two sections that customize `dev` profile configuration (doing so even in a related way) would exacerbate the current unclear situation. I'd like to remove whatever is not planned or adjust comments to reflect the goal of code that is wanted for the future but not yet, and then if applicable put the most related things together.

### Expected behavior 🤔

See above. It seemed clearer to include this for each subsection.

(I considered opening multiple issues or opening no issue for anything here and just bringing it all up in comments, but I figured this might still be best, even though it goes a bit against the format, because there is at least one other issue related to the comments there, #1478, that should be covered separately and should not be confused with any of this.)

### Git behavior

I have not researched this. Some choices of optimization level or the treatment of debug symbols might be illuminating. But the tooling and language are different, and for the upstream Git project, I believe there are no official binary releases. So it seems largely inapplicable.

### Steps to reproduce 🕹

Since this is a clarity and release-engineering issue opened for tracking and feedback about things I think are mostly known, reproduction consists mainly of examining the code shown, linked, or otherwise discussed above and in the related portion of the #1475 comment discussion.

One practical note, though: On some platforms, the `file` command will reveal debug symbols, including text like "not stripped" in its output when they are present. But it seems the `file` command on macOS may not do this.

### PR DESCRIPTION
**Pull request by EliahKagan**
This makes the kind of changes discussed in comments in #1475 and further in #1477. This fixes #1477, though some other fix might be preferred, so this may benefit from being changed, or further changes could be made afterward. See https://github.com/Byron/gitoxide/issues/1477#issuecomment-2257499363 and https://github.com/Byron/gitoxide/issues/1477#issuecomment-2274445001.

In `Cargo.toml`, this adds thin LTO to the `release` profile, and creates a separate more aggressively optimized `release-opt` profile with fat LTO and max 1 codegen unit. The effect on build times seems in both cases modest, but you might disagree. For more on that, see [this gist where I timed the builds](https://gist.github.com/EliahKagan/52cedc99a8544505525706351948664a) and https://github.com/Byron/gitoxide/issues/1477#issuecomment-2274445001 (which, among other related topics, discusses it). This also removes some old commented-out code, as also discussed in #1477.

In `release.yml`, this takes the profile to use from an environment variable (some occurrences of `release` had referred to that profile while others had not), does some related other refactoring, and most significantly changes which profile is used for the GitHub releases from `release` to the new `release-opt` profile.

The results [seem okay](https://gist.github.com/EliahKagan/a8a59e0f6fb4435dcbd9555b3c8b8680), and in particular the Linux executables, which without configuring stripping in `Cargo.toml` might require more explicit logic as new targets are added, show as stripped. (The `file` implementation I am using does not seem to report the absence of symbols in macOS binaries; that affected prior runs' results as well.) That output is on assets downloaded from a draft release produced in [this workflow run](https://github.com/EliahKagan/gitoxide/actions/runs/10293234759).

In case you decide to merge this without many changes but hope for an even better approach later, I've [included a comment](https://github.com/Byron/gitoxide/pull/1495/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R216-R217) in `Cargo.toml` stating that the `release-opt` profile is not guaranteed to stick around.

### CODE REVIEW

File: .github/workflows/release.yml
Comment (EliahKagan) [inline .github/workflows/release.yml:337]: Before merging, actual publishing should be reenabled:

```suggestion
      - name: Publish the release
```

I have no objection to this being merged at any time with that change, but as noted above I suspect you may want other changes, so for now this is still marked as a draft.

Comment (EliahKagan) [reply .github/workflows/release.yml]: Obsolete,  871eec7 includes this.

File: Cargo.toml
Comment (EliahKagan) [inline Cargo.toml:209]: It seems to me that `lto = "thin"` here in the `release` profile is reasonable, but as noted above it does slow down builds a bit. In case you don't want it, you can accept this suggestion:

```suggestion
```

This is for convenience if that is the case, but not to advocate this change.

It might make sense to wait until benchmarking is done before adding this, though I believe it usually improves runtime performance--unlike going from thin to fat LTO, which is not always the case--and decisively determining that it is not worthwhile may involve benchmarking on quite a wide variety of architectures.

(This change would not affect the `release-opt` profile, because while that inherits from this, it has its own value here, setting it to `fat`.)

**Review by Byron — APPROVED** [review]
Thanks a lot for proposing the change and for acquiring numbers!
I think going to LTO="thin" should be fine and maybe it's good for a speedboost at some point, even though maybe one day we might go back to the previous setting if it turns out not to be the case.
Out of ~~caution~~ fear I also removed the opt-level=0 statement to rather leave it at the default.

Finally, I renamed the profile to clearly state what it is for, `-github`, which I think should make it less likely to be used by others - `-opt` sounded general enough for usage by others who want an even more optimized build.

Please do double-check my changes in case they have a chance to break something, after all I didn't really test any of this.

With the current settings, on my M1 Pro a release build can be done in 58s, which seems fine. A `release-github` build takes 2m30s, for comparison.

As a side-note: The `inherits` directive is required even though I tried not to use it, so instead I repeated all values from the `release` profile to assure they are not actually inherited, but rather, controlled explicitly.

File: Cargo.toml
Comment (EliahKagan) [inline Cargo.toml:215]: I'm not sure I understand the reason for removing this based on the information in the commit message, though I can guess a possible good reason. This was added in add2e3e (#2) and only reduces the optimization level for `build` dependencies, i.e., those that are used for building. So the resulting executables' level of optimization is unaffected by this. But a note in [the documentation](https://doc.rust-lang.org/cargo/reference/profiles.html#overrides) mentions:

> When a dependency is both a normal dependency and a build dependency, Cargo will try to only build it once when `--target` is not specified. When using `build-override`, the dependency may need to be built twice, once as a normal dependency and once with the overridden build settings. This may increase initial build times.

Is this the reason for removing it?

File: Cargo.toml
Comment (EliahKagan) [inline Cargo.toml:224]: Assuming `build-override` is no longer to be set in the `release` profile from which this one inherits, isn't this superfluous? In the absence of `build-override`, I think build dependencies have the same `opt-level` as the profile, and since that is `release` here, we already have `opt-level = 3`.

You mentioned:

> As a side-note: The `inherits` directive is required even though I tried not to use it, so instead I repeated all values from the `release` profile to assure they are not actually inherited, but rather, controlled explicitly.

But only the values changed in `Cargo.toml` were repeated, except this, which is explicit here but implicit in the `release` profile. (`opt-level` for the profiles themselves is implicit in both.)

**Review by EliahKagan — COMMENTED** [review]
> Out of ~~caution~~ fear I also removed the opt-level=0 statement to rather leave it at the default.

I've requested clarification about this in my [first review comment](https://github.com/Byron/gitoxide/pull/1495/files#r1708741138) below. The presence of `build-override` with a lower `opt-level` can affect build times, usually decreasing them unless it causes the same crates to be built more than once and otherwise sometimes increasing them. It should not affect the runtime performance of the `gix` and `ein` binaries, regardless of whether and how it is set.

> Finally, I renamed the profile to clearly state what it is for

Good idea!

> Please do double-check my changes in case they have a chance to break something, after all I didn't really test any of this.

I haven't rerun the workflow yet, but on inspection it doesn't seem like anything would be broken.

> As a side-note: The `inherits` directive *[...}*

This relates to my [second review comment](https://github.com/Byron/gitoxide/pull/1495/files#r1708752857) below. I'm not sure the current amount of repetition is clearer, since it is only repeating what `release` overrides explicitly from the `cargo` defaults--quite a bit remains implicit in both profile definitions.

It doesn't seem particularly unclear though, except with respect to the intent of one change that the second comment concerns.

File: Cargo.toml
Comment (Byron) [reply Cargo.toml:215]: I completely misunderstood `build-override` and was thinking it's related to an opt-level override for anything that is built, not just build scripts.

Avoiding to possibly build dependencies twice It's probably a good reason to remove it, especially given that there are no build dependencies with a build-script that just uses the standard library.

File: Cargo.toml
Comment (Byron) [reply Cargo.toml:224]: With my recent gain of knowledge, I think this line should be removed. This was done in `main` now.

**Comment by EliahKagan** [comment]
> > Please do double-check my changes in case they have a chance to break something, after all I didn't really test any of this.
>
> I haven't rerun the workflow yet, but on inspection it doesn't seem like anything would be broken.

I have actually [tested this](https://github.com/EliahKagan/gitoxide/actions/runs/10345803160) now, and [it seems fine](https://gist.github.com/EliahKagan/12398ba669c19cf1f94444969d0d55f0).

**Pull request merged**
